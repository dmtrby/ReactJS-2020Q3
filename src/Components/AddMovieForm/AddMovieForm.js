import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Input from '../Base/Input';
import Button from '../Base/Button';

import './movie-form.scss';

const AddMovieForm = ({ name, year, url, genres, description, duration, id, onSubmit, hideModal }) => {
  const [form, setForm] = useState({
    id,
    name,
    year,
    url,
    genres,
    description,
    duration,
  });

  const handleInputChange = (e) => {
    const { target } = e;
    const { value, name: tName } = target;
    if (tName === 'genres') {
      setForm({
        ...form,
        [tName]: value.split(' '),
      });
    } else {
      setForm({
        ...form,
        [tName]: value,
      });
    }
  };

  const resetForm = () => {
    setForm({
      name: '',
      year: '',
      url: '',
      genres: [],
      description: '',
      duration: '',
    });
  };

  const submitHandler = (e) => {
    e.preventDefault();
    onSubmit({ ...form });
    hideModal();
  };

  let genresString = form.genres;
  if (Array.isArray(genres)) {
    genresString = form.genres.join(' ');
  }
  return (
    <form onSubmit={submitHandler} className="movie-form container">
      <div className="row">
        {form.id && (
          <div className="col-xs-12">
            <Input type="text" disabled value={id} labelText="Id" id="id" name="id" placeHolder="Movie Id" />
          </div>
        )}
        <div className="col-xs-12">
          <Input
            type="text"
            value={form.name}
            onChange={handleInputChange}
            labelText="Title"
            id="name"
            name="name"
            placeHolder="Select title"
          />
        </div>
        <div className="col-xs-12">
          <Input
            type="text"
            value={form.year}
            onChange={handleInputChange}
            labelText="Release date"
            id="year"
            name="year"
            placeHolder="Select date (year)"
          />
        </div>
        <div className="col-xs-12">
          <Input
            type="text"
            value={form.url}
            onChange={handleInputChange}
            labelText="Movie url"
            id="url"
            name="url"
            placeHolder="Movie URL here"
          />
        </div>
        <div className="col-xs-12">
          <Input
            type="text"
            value={genresString}
            onChange={handleInputChange}
            labelText="Genres"
            id="genres"
            name="genres"
            placeHolder="Select Genres"
          />
        </div>
        <div className="col-xs-12">
          <Input
            type="text"
            value={form.description}
            onChange={handleInputChange}
            labelText="Description"
            id="description"
            name="description"
            placeHolder="Description here"
          />
        </div>
        <div className="col-xs-12">
          <Input
            type="text"
            value={form.duration}
            onChange={handleInputChange}
            labelText="Duration"
            id="duration"
            name="duration"
            placeHolder="Duration here"
          />
        </div>
        <div className="col-xs-12 margin-top-4">
          <div className="row end-xs">
            <Button
              type="reset"
              onClick={resetForm}
              text="reset"
              variant="secondary"
              classList="margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
            />
            <Button
              type="submit"
              text="submit"
              variant="primary"
              classList="margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
            />
          </div>
        </div>
      </div>
    </form>
  );
};

AddMovieForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  hideModal: PropTypes.func,
  name: PropTypes.string,
  year: PropTypes.string,
  url: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  genres: PropTypes.array,
  description: PropTypes.string,
  duration: PropTypes.string,
  id: PropTypes.string,
};

AddMovieForm.defaultProps = {
  name: '',
  year: '',
  url: '',
  genres: [],
  description: '',
  duration: '',
  id: '',
  hideModal: () => console.warn('hide modal func doesnt exist'),
};

export default AddMovieForm;
