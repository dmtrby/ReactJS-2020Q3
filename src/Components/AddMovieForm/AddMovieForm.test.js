import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';

import AddMovieForm from './AddMovieForm';

const onSubmitFn = jest.fn();

describe('AddMovieForm visual for add new movie component', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
    render(<AddMovieForm onSubmit={onSubmitFn} />);
  });

  it('renders input elements', () => {
    expect(screen.queryByText('id')).toBeNull();
    expect(screen.queryByText(/title/i)).toBeTruthy();
    expect(screen.queryByText(/release date/i)).toBeTruthy();
    expect(screen.queryByText(/poster path/i)).toBeTruthy();
    expect(screen.queryByText(/genres/i)).toBeTruthy();
    expect(screen.queryByText(/overview/i)).toBeTruthy();
    expect(screen.queryByText(/runtime/i)).toBeTruthy();
  });

  it('renders buttons ', () => {
    expect(screen.queryByText(/submit/i)).toBeTruthy();
    expect(screen.queryByText(/reset/i)).toBeTruthy();
  });
});

describe('AddMovieForm visual for edit movie component', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
    render(<AddMovieForm id={1} genres={['horror', 'comedy']} onSubmit={onSubmitFn} />);
  });

  it('renders input elements', () => {
    expect(screen.queryByText(/id/i)).toBeTruthy();
    expect(screen.queryByText(/title/i)).toBeTruthy();
    expect(screen.queryByText(/release date/i)).toBeTruthy();
    expect(screen.queryByText(/poster path/i)).toBeTruthy();
    expect(screen.queryByText(/genres/i)).toBeTruthy();
    expect(screen.queryByText(/overview/i)).toBeTruthy();
    expect(screen.queryByText(/runtime/i)).toBeTruthy();
  });

  it('renders buttons ', () => {
    expect(screen.queryByText(/submit/i)).toBeTruthy();
    expect(screen.queryByText(/reset/i)).toBeTruthy();
  });
});

describe('Inputs', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
    render(<AddMovieForm id={1} genres={['horror', 'comedy']} onSubmit={onSubmitFn} />);
  });

  it('Should show input validation after blur and hide if validation is correct for title', async () => {
    const input = screen.getByLabelText('title');
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.blur(input);

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/required/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });

    fireEvent.change(input, { target: { value: 'm' } });

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/Should be more than 2 symbols/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });

    fireEvent.change(input, { target: { value: 'mocked title' } });

    await waitFor(() => {
      expect(input.classList.contains('input-field--valid')).toBe(true);
      expect(input.classList.contains('input-field--invalid')).toBe(false);
      const error = screen.queryByText(/Should be more than 2 symbols/i);
      expect(error).toBe(null);
    });
  });

  it('Should show input validation after blur and hide if validation is correct for release date', async () => {
    const input = screen.getByLabelText(/release date/i);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.blur(input);

    await waitFor(() => {
      expect(input.classList.contains('input-field--valid')).toBe(false);
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/required/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });

    fireEvent.change(input, { target: { value: '2011-02-03' } });

    await waitFor(() => input.classList.contains('input-field--valid'));

    expect(input.classList.contains('input-field--valid')).toBe(true);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    const error = screen.queryByText(/required/i);
    expect(error).toBe(null);
  });

  it('Should show input validation after blur and hide if validation is correct for poster path', async () => {
    const input = screen.getByLabelText(/poster path/i);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.blur(input);

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/required/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });
  });

  it('Should show input validation after blur and hide if validation is correct for overview', async () => {
    const input = screen.getByLabelText(/overview/i);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.blur(input);

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/required/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });
  });

  it('Should show input validation after blur and hide if validation is correct for runtime', async () => {
    const input = screen.getByLabelText(/runtime/i);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.blur(input);

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryByText(/required/i);
      expect(error.classList.contains('input-field__error')).toBeTruthy();
    });
  });

  it('Should show validation of all inputs after submit', async () => {
    const input = screen.getByLabelText('title');
    const submitButton = screen.queryByText(/submit/i);
    expect(input.classList.contains('input-field--invalid')).toBe(false);
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(input.classList.contains('input-field--invalid')).toBe(true);
      const error = screen.queryAllByText(/required/i);
      expect(error).toBeTruthy();
      expect(onSubmitFn).toHaveBeenCalledTimes(0);
    });
  });

  it('Should call onSubmit if there are no validations errors', async () => {
    fireEvent.change(screen.queryByLabelText(/title/i), { target: { value: 'mocked title' } });
    fireEvent.change(screen.queryByLabelText(/release date/i), { target: { value: '2011-02-03' } });
    fireEvent.change(screen.queryByLabelText(/poster path/i), { target: { value: 'http://google.com' } });
    fireEvent.change(screen.queryByLabelText(/overview/i), { target: { value: 'mocked overview' } });
    fireEvent.change(screen.queryByLabelText(/runtime/i), { target: { value: '130' } });

    

    const submitButton = screen.queryByText(/submit/i);
    fireEvent.click(submitButton);

    expect(onSubmitFn).toHaveBeenCalledTimes(0); // todo
  });
});

describe('AddMovieForm Snapshots', () => {
  test('AddMovieForm matches the Snapshot', () => {
    const component = renderer.create(<AddMovieForm onSubmit={onSubmitFn} />);
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('EditMovieForm matches the Snapshot', () => {
    const component = renderer.create(<AddMovieForm id={1} onSubmit={onSubmitFn} />);
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});



