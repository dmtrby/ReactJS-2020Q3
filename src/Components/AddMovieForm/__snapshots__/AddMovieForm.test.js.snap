// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieForm Snapshots AddMovieForm matches the Snapshot 1`] = `
<form
  action="#"
  autoComplete="off"
  className="movie-form container"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="title"
        >
          title
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="title"
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Select title"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="release_date"
        >
          Release date
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="release_date"
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Select release date"
          type="date"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="poster_path"
        >
          Poster path
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="poster_path"
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Poster URL here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <div
          className="col-xs-12 padding-0"
        >
          <label
            className="input-field__label col-xs-12"
            htmlFor="genres"
          >
            Genres
          </label>
          <div
            className="select css-2b097c-container"
            id="genres"
            onKeyDown={[Function]}
          >
            <div
              className="select__control css-10q3x8k-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className="select__value-container select__value-container--is-multi css-g1d714-ValueContainer"
              >
                <div
                  className="select__placeholder css-1wa3eu0-placeholder"
                >
                  Select...
                </div>
                <div
                  className="css-b8ldur-Input"
                >
                  <div
                    className="select__input"
                    style={
                      Object {
                        "display": "inline-block",
                      }
                    }
                  >
                    <input
                      aria-autocomplete="list"
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect="off"
                      disabled={false}
                      id="react-select-13-input"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      spellCheck="false"
                      style={
                        Object {
                          "background": 0,
                          "border": 0,
                          "boxSizing": "content-box",
                          "color": "inherit",
                          "fontSize": "inherit",
                          "label": "input",
                          "opacity": 1,
                          "outline": 0,
                          "padding": 0,
                          "width": "1px",
                        }
                      }
                      tabIndex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style={
                        Object {
                          "height": 0,
                          "left": 0,
                          "overflow": "scroll",
                          "position": "absolute",
                          "top": 0,
                          "visibility": "hidden",
                          "whiteSpace": "pre",
                        }
                      }
                    >
                      
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  className="select__indicator-separator css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-6q0nyr-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="overview"
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Overview here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="runtime"
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Runtime here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12 margin-top-4"
    >
      <div
        className="row end-xs"
      >
        <button
          className="button button--secondary  margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
          onClick={[Function]}
          type="reset"
        >
          reset
        </button>
        <button
          className="button button--primary  margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
          onClick={[Function]}
          type="submit"
        >
          submit
        </button>
      </div>
    </div>
  </div>
</form>
`;

exports[`AddMovieForm Snapshots EditMovieForm matches the Snapshot 1`] = `
<form
  action="#"
  autoComplete="off"
  className="movie-form container"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="id"
        >
          Id
        </label>
        <input
          className="input-field__label__input col-xs-12 "
          disabled={true}
          id="id"
          name=""
          onChange={[Function]}
          placeholder=""
          type="text"
          value={1}
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="title"
        >
          title
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="title"
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Select title"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="release_date"
        >
          Release date
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="release_date"
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Select release date"
          type="date"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="poster_path"
        >
          Poster path
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="poster_path"
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Poster URL here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <div
          className="col-xs-12 padding-0"
        >
          <label
            className="input-field__label col-xs-12"
            htmlFor="genres"
          >
            Genres
          </label>
          <div
            className="select css-2b097c-container"
            id="genres"
            onKeyDown={[Function]}
          >
            <div
              className="select__control css-10q3x8k-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className="select__value-container select__value-container--is-multi css-g1d714-ValueContainer"
              >
                <div
                  className="select__placeholder css-1wa3eu0-placeholder"
                >
                  Select...
                </div>
                <div
                  className="css-b8ldur-Input"
                >
                  <div
                    className="select__input"
                    style={
                      Object {
                        "display": "inline-block",
                      }
                    }
                  >
                    <input
                      aria-autocomplete="list"
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect="off"
                      disabled={false}
                      id="react-select-14-input"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      spellCheck="false"
                      style={
                        Object {
                          "background": 0,
                          "border": 0,
                          "boxSizing": "content-box",
                          "color": "inherit",
                          "fontSize": "inherit",
                          "label": "input",
                          "opacity": 1,
                          "outline": 0,
                          "padding": 0,
                          "width": "1px",
                        }
                      }
                      tabIndex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style={
                        Object {
                          "height": 0,
                          "left": 0,
                          "overflow": "scroll",
                          "position": "absolute",
                          "top": 0,
                          "visibility": "hidden",
                          "whiteSpace": "pre",
                        }
                      }
                    >
                      
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  className="select__indicator-separator css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-6q0nyr-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="overview"
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Overview here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="input-field row"
      >
        <label
          className="input-field__label col-xs-12"
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          className="input-field__label__input col-xs-12  "
          id="runtime"
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Runtime here"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col-xs-12 margin-top-4"
    >
      <div
        className="row end-xs"
      >
        <button
          className="button button--secondary  margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
          onClick={[Function]}
          type="reset"
        >
          reset
        </button>
        <button
          className="button button--primary  margin-left-lg-2 margin-top-lg-0 margin-top-xs-1"
          onClick={[Function]}
          type="submit"
        >
          submit
        </button>
      </div>
    </div>
  </div>
</form>
`;
